File: products/compliance/src/analyzers/git-analyzer.ts
import type { Analyzer, Finding, RuleDefinition, ScannedFile } from '../types.js';
import { readFileSync } from 'node:fs';
import { basename } from 'node:path';

/**
 * Analyzes repository hygiene for compliance issues.
 *
 * Checks include (static file inspection only - no git commands):
 * - .gitignore completeness (*.pem, *.key, .env entries)
 * - CODEOWNERS file presence
 * - SECURITY.md file presence
 * - LICENSE file presence
 * - Large files (>5MB) that should not be in git
 */
export class GitAnalyzer implements Analyzer {
  name = 'git-analyzer';

  private static readonly LARGE_FILE_THRESHOLD = 5 * 1024 * 1024; // 5 MB

  async analyze(files: ScannedFile[], rules: RuleDefinition[]): Promise<Finding[]> {
    const myRules = rules.filter((r) => r.analyzer === this.name);
    if (myRules.length === 0) return [];

    // Build lookup from pattern -> rules (multiple rules may share a pattern)
    const rulesByPattern = new Map<string, RuleDefinition[]>();
    for (const rule of myRules) {
      const existing = rulesByPattern.get(rule.pattern) ?? [];
      existing.push(rule);
      rulesByPattern.set(rule.pattern, existing);
    }

    const findings: Finding[] = [];

    // ── .gitignore completeness ─────────────────────────────────────
    this.checkGitignoreCompleteness(files, rulesByPattern, findings);

    // ── CODEOWNERS ──────────────────────────────────────────────────
    this.checkCodeowners(files, rulesByPattern, findings);

    // ── SECURITY.md ─────────────────────────────────────────────────
    this.checkSecurityPolicy(files, rulesByPattern, findings);

    // ── LICENSE ─────────────────────────────────────────────────────
    this.checkLicense(files, rulesByPattern, findings);

    // ── Large files ─────────────────────────────────────────────────
    this.checkLargeFiles(files, rulesByPattern, findings);

    return findings;
  }

  /**
   * Check .gitignore files for completeness:
   * - .env entry
   * - *.pem and *.key entries
   * - General completeness
   */
  private checkGitignoreCompleteness(
    files: ScannedFile[],
    rulesByPattern: Map<string, RuleDefinition[]>,
    findings: Finding[],
  ): void {
    const gitignoreFiles = files.filter((f) => basename(f.path) === '.gitignore');

    for (const gitignoreFile of gitignoreFiles) {
      let content: string;
      try {
        content = readFileSync(gitignoreFile.path, 'utf-8');
      } catch {
        continue;
      }

      const lines = content.split('\n').map((l) => l.trim());

      // Check for .env
      const hasEnv = lines.some(
        (line) => line === '.env' || line === '.env*' || line === '*.env',
      );
      if (!hasEnv) {
        this.addFindings(findings, 'env-not-gitignored', rulesByPattern, {
          file: gitignoreFile.relativePath,
          evidence: '.env not listed in .gitignore',
        });
      }

      // Check for *.pem and *.key
      const hasPem = lines.some((line) => line === '*.pem');
      const hasKey = lines.some((line) => line === '*.key');
      if (!hasPem || !hasKey) {
        const missing = [];
        if (!hasPem) missing.push('*.pem');
        if (!hasKey) missing.push('*.key');
        this.addFindings(findings, 'sensitive-not-gitignored', rulesByPattern, {
          file: gitignoreFile.relativePath,
          evidence: `Missing gitignore entries: ${missing.join(', ')}`,
        });
      }

      // General incompleteness: flag if missing .env OR (*.pem AND *.key)
      if (!hasEnv || !hasPem || !hasKey) {
        this.addFindings(findings, 'incomplete-gitignore', rulesByPattern, {
          file: gitignoreFile.relativePath,
          evidence: 'Gitignore is missing recommended entries for sensitive files',
        });
      }
    }
  }

  /**
   * Check for CODEOWNERS file in standard locations:
   * CODEOWNERS, .github/CODEOWNERS, docs/CODEOWNERS
   */
  private checkCodeowners(
    files: ScannedFile[],
    rulesByPattern: Map<string, RuleDefinition[]>,
    findings: Finding[],
  ): void {
    const hasCodeowners = files.some((f) => {
      const rel = f.relativePath;
      return (
        rel === 'CODEOWNERS' ||
        rel === '.github/CODEOWNERS' ||
        rel === 'docs/CODEOWNERS'
      );
    });

    if (!hasCodeowners) {
      this.addFindings(findings, 'missing-codeowners', rulesByPattern, {
        evidence: 'No CODEOWNERS file found in repository',
      });
    }
  }

  /**
   * Check for SECURITY.md file.
   */
  private checkSecurityPolicy(
    files: ScannedFile[],
    rulesByPattern: Map<string, RuleDefinition[]>,
    findings: Finding[],
  ): void {
    const hasSecurityPolicy = files.some(
      (f) => basename(f.path).toUpperCase() === 'SECURITY.MD',
    );

    if (!hasSecurityPolicy) {
      this.addFindings(findings, 'no-security-policy', rulesByPattern, {
        evidence: 'No SECURITY.md file found in repository',
      });
    }
  }

  /**
   * Check for LICENSE file (LICENSE, LICENSE.md, or LICENSE.txt).
   */
  private checkLicense(
    files: ScannedFile[],
    rulesByPattern: Map<string, RuleDefinition[]>,
    findings: Finding[],
  ): void {
    const hasLicense = files.some((f) => {
      const name = basename(f.path).toUpperCase();
      return name === 'LICENSE' || name === 'LICENSE.MD' || name === 'LICENSE.TXT';
    });

    if (!hasLicense) {
      this.addFindings(findings, 'missing-license', rulesByPattern, {
        evidence: 'No LICENSE file found in repository',
      });
    }
  }

  /**
   * Flag any file larger than 5MB.
   */
  private checkLargeFiles(
    files: ScannedFile[],
    rulesByPattern: Map<string, RuleDefinition[]>,
    findings: Finding[],
  ): void {
    for (const file of files) {
      if (file.size > GitAnalyzer.LARGE_FILE_THRESHOLD) {
        const sizeMB = (file.size / (1024 * 1024)).toFixed(1);
        this.addFindings(findings, 'large-binary-in-git', rulesByPattern, {
          file: file.relativePath,
          evidence: `File size: ${sizeMB} MB (threshold: 5 MB)`,
        });
      }
    }
  }

  /**
   * Create findings for all rules matching a given pattern.
   * A single pattern may map to rules in multiple frameworks (soc2, hipaa, gdpr).
   */
  private addFindings(
    findings: Finding[],
    pattern: string,
    rulesByPattern: Map<string, RuleDefinition[]>,
    context: { file?: string; line?: number; evidence?: string },
  ): void {
    const matchedRules = rulesByPattern.get(pattern);
    if (!matchedRules || matchedRules.length === 0) return;

    for (const rule of matchedRules) {
      findings.push({
        id: rule.id,
        title: rule.title,
        description: rule.description,
        severity: rule.severity,
        framework: rule.framework,
        controlIds: rule.controls,
        file: context.file,
        line: context.line,
        evidence: context.evidence,
        analyzer: this.name,
        fixable: rule.fixable,
      });
    }
  }
}

File: products/compliance-go/analyzers/git_analyzer.go
package analyzers

import (
	"os"
	"path/filepath"
	"strings"

	"github.com/rishav1305/soul/products/compliance-go/rules"
)

// GitAnalyzer implements the Analyzer interface for detecting project-level
// git hygiene issues: missing .gitignore entries, CODEOWNERS, SECURITY.md,
// LICENSE, and CI configuration.
type GitAnalyzer struct{}

// Name returns the analyzer identifier used to match rules.
func (g *GitAnalyzer) Name() string {
	return "git-analyzer"
}

// Analyze checks project-level git configuration and file presence.
func (g *GitAnalyzer) Analyze(files []ScannedFile, allRules []rules.Rule) ([]Finding, error) {
	myRules := filterRules(allRules, g.Name())
	if len(myRules) == 0 {
		return nil, nil
	}

	rulesByPattern := groupRulesByPattern(myRules)
	var findings []Finding

	// Build a set of relative paths for quick lookups.
	pathSet := make(map[string]ScannedFile)
	for _, f := range files {
		pathSet[f.RelativePath] = f
	}

	// Check .gitignore for missing entries
	findings = append(findings, g.checkGitignore(pathSet, rulesByPattern)...)

	// Check for CODEOWNERS
	findings = append(findings, g.checkFileExists(pathSet, rulesByPattern,
		"missing-codeowners",
		[]string{"CODEOWNERS", ".github/CODEOWNERS", "docs/CODEOWNERS"},
		"No CODEOWNERS file found",
	)...)

	// Check for SECURITY.md
	findings = append(findings, g.checkFileExists(pathSet, rulesByPattern,
		"no-security-policy",
		[]string{"SECURITY.md", ".github/SECURITY.md"},
		"No SECURITY.md file found",
	)...)

	// Check for LICENSE
	findings = append(findings, g.checkFileExists(pathSet, rulesByPattern,
		"missing-license",
		[]string{"LICENSE", "LICENSE.md", "LICENSE.txt", "LICENCE", "LICENCE.md", "LICENCE.txt"},
		"No LICENSE file found",
	)...)

	return findings, nil
}

// checkGitignore checks if .gitignore exists and contains important entries.
func (g *GitAnalyzer) checkGitignore(pathSet map[string]ScannedFile, rulesByPattern map[string][]rules.Rule) []Finding {
	var findings []Finding

	gi, ok := pathSet[".gitignore"]
	if !ok {
		// .gitignore itself is missing — report as incomplete
		for _, rule := range rulesByPattern["incomplete-gitignore"] {
			findings = append(findings, Finding{
				ID:          rule.ID,
				Title:       rule.Title,
				Description: rule.Description,
				Severity:    rule.Severity,
				Framework:   rule.Framework,
				ControlIDs:  rule.Controls,
				Evidence:    "No .gitignore file found in the repository",
				Analyzer:    g.Name(),
				Fixable:     rule.Fixable,
			})
		}
		return findings
	}

	content, err := os.ReadFile(gi.Path)
	if err != nil {
		return nil
	}

	gitignoreContent := string(content)
	lines := strings.Split(gitignoreContent, "\n")

	// Check for .env entry
	hasEnv := false
	for _, line := range lines {
		trimmed := strings.TrimSpace(line)
		if trimmed == ".env" || trimmed == ".env*" || trimmed == ".env.*" {
			hasEnv = true
			break
		}
	}
	if !hasEnv {
		for _, rule := range rulesByPattern["env-not-gitignored"] {
			findings = append(findings, Finding{
				ID:          rule.ID,
				Title:       rule.Title,
				Description: rule.Description,
				Severity:    rule.Severity,
				Framework:   rule.Framework,
				ControlIDs:  rule.Controls,
				File:        ".gitignore",
				Evidence:    ".gitignore does not contain .env entry",
				Analyzer:    g.Name(),
				Fixable:     rule.Fixable,
			})
		}
	}

	// Check for node_modules entry
	hasNodeModules := false
	for _, line := range lines {
		trimmed := strings.TrimSpace(line)
		if trimmed == "node_modules" || trimmed == "node_modules/" {
			hasNodeModules = true
			break
		}
	}

	// Check for common secret patterns (*.pem, *.key, etc.)
	sensitivePatterns := []string{"*.pem", "*.key", "*.p12", "*.pfx", "credentials.json"}
	missingSensitive := []string{}
	for _, pat := range sensitivePatterns {
		found := false
		for _, line := range lines {
			trimmed := strings.TrimSpace(line)
			if trimmed == pat {
				found = true
				break
			}
		}
		if !found {
			missingSensitive = append(missingSensitive, pat)
		}
	}

	if len(missingSensitive) > 0 {
		for _, rule := range rulesByPattern["sensitive-not-gitignored"] {
			findings = append(findings, Finding{
				ID:          rule.ID,
				Title:       rule.Title,
				Description: rule.Description,
				Severity:    rule.Severity,
				Framework:   rule.Framework,
				ControlIDs:  rule.Controls,
				File:        ".gitignore",
				Evidence:    "Missing entries: " + strings.Join(missingSensitive, ", "),
				Analyzer:    g.Name(),
				Fixable:     rule.Fixable,
			})
		}
	}

	// Check for incomplete .gitignore (missing node_modules or other common entries)
	if !hasNodeModules {
		for _, rule := range rulesByPattern["incomplete-gitignore"] {
			findings = append(findings, Finding{
				ID:          rule.ID,
				Title:       rule.Title,
				Description: rule.Description,
				Severity:    rule.Severity,
				Framework:   rule.Framework,
				ControlIDs:  rule.Controls,
				File:        ".gitignore",
				Evidence:    ".gitignore is missing node_modules entry",
				Analyzer:    g.Name(),
				Fixable:     rule.Fixable,
			})
		}
	}

	return findings
}

// checkFileExists checks if at least one of the given file paths exists.
// If none exist, findings are created using the given pattern.
func (g *GitAnalyzer) checkFileExists(pathSet map[string]ScannedFile, rulesByPattern map[string][]rules.Rule, pattern string, paths []string, evidence string) []Finding {
	for _, p := range paths {
		// Check both exact match and case-insensitive
		if _, ok := pathSet[p]; ok {
			return nil
		}
	}

	// Also check by basename (case-insensitive) for flexibility
	for _, f := range pathSet {
		base := filepath.Base(f.RelativePath)
		for _, p := range paths {
			if strings.EqualFold(base, filepath.Base(p)) {
				return nil
			}
		}
	}

	var findings []Finding
	for _, rule := range rulesByPattern[pattern] {
		findings = append(findings, Finding{
			ID:          rule.ID,
			Title:       rule.Title,
			Description: rule.Description,
			Severity:    rule.Severity,
			Framework:   rule.Framework,
			ControlIDs:  rule.Controls,
			Evidence:    evidence,
			Analyzer:    g.Name(),
			Fixable:     rule.Fixable,
		})
	}
	return findings
}
